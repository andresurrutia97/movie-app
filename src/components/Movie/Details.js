import React from 'react';
import PropTypes from 'prop-types';

const InfoDetails = ({ title, name }) => {
  return (
    <div className="flex flex-col items-center mx-2 md:flex-row md:mx-0 md:mr-2">
      <span className="text-sm text-white">{title}</span>
      <span className="hidden md:block md:mr-2">:</span>
      <span className="font-semibold text-white md:mr-2">{name}</span>
    </div>
  );
};

InfoDetails.propTypes = {
  title: PropTypes.string.isRequired,
  name: PropTypes.string,
};

InfoDetails.defaultProps = {
  name: '',
};
export default InfoDetails;
