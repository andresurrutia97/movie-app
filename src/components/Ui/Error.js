import React from 'react';

const Error = () => (
  <p className="w-full py-5 text-center">Something went wrong</p>
);

export default Error;
