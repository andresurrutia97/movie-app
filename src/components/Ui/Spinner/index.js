import React from 'react';

import './index.css';

const Loading = () => <div className="spinner" />;

export default Loading;
