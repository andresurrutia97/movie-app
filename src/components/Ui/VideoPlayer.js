import React from 'react';
import PropTypes from 'prop-types';
import ReactPlayer from 'react-player';

const VideoPlayer = ({ url }) => {
  return (
    <div className="h-full shadow-md">
      <ReactPlayer url={url} width="100%" height="100%" controls playing />
    </div>
  );
};

VideoPlayer.propTypes = {
  url: PropTypes.string.isRequired,
};

export default VideoPlayer;
