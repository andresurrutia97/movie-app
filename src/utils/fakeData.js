export const moviesData = {
  page: 1,
  results: [
    {
      poster_path: '/e1mjopzAS2KNsvpbpahQ1a6SkSn.jpg',
      adult: false,
      overview:
        'From DC Comics comes the Suicide Squad, an antihero team of incarcerated supervillains who act as deniable assets for the United States government, undertaking high-risk black ops missions in exchange for commuted prison sentences.',
      release_date: '2016-08-03',
      genre_ids: [14, 28, 80],
      id: 297761,
      original_title: 'Suicide Squad',
      original_language: 'en',
      title: 'Suicide Squad',
      backdrop_path: '/ndlQ2Cuc3cjTL7lTynw6I4boP4S.jpg',
      popularity: 48.261451,
      vote_count: 1466,
      video: false,
      vote_average: 5.91,
    },
    {
      poster_path: '/lFSSLTlFozwpaGlO31OoUeirBgQ.jpg',
      adult: false,
      overview:
        'The most dangerous former operative of the CIA is drawn out of hiding to uncover hidden truths about his past.',
      release_date: '2016-07-27',
      genre_ids: [28, 53],
      id: 324668,
      original_title: 'Jason Bourne',
      original_language: 'en',
      title: 'Jason Bourne',
      backdrop_path: '/AoT2YrJUJlg5vKE3iMOLvHlTd3m.jpg',
      popularity: 30.690177,
      vote_count: 649,
      video: false,
      vote_average: 5.25,
    },
    {
      poster_path: '/hU0E130tsGdsYa4K9lc3Xrn5Wyt.jpg',
      adult: false,
      overview:
        'One year after outwitting the FBI and winning the publicâ€™s adulation with their mind-bending spectacles, the Four Horsemen resurface only to find themselves face to face with a new enemy who enlists them to pull off their most dangerous heist yet.',
      release_date: '2016-06-02',
      genre_ids: [28, 12, 35, 80, 9648, 53],
      id: 291805,
      original_title: 'Now You See Me 2',
      original_language: 'en',
      title: 'Now You See Me 2',
      backdrop_path: '/zrAO2OOa6s6dQMQ7zsUbDyIBrAP.jpg',
      popularity: 29.737342,
      vote_count: 684,
      video: false,
      vote_average: 6.64,
    },
    {
      poster_path: '/h28t2JNNGrZx0fIuAw8aHQFhIxR.jpg',
      adult: false,
      overview:
        'A recently cheated on married woman falls for a younger man who has moved in next door, but their torrid affair soon takes a dangerous turn.',
      release_date: '2015-01-23',
      genre_ids: [53],
      id: 241251,
      original_title: 'The Boy Next Door',
      original_language: 'en',
      title: 'The Boy Next Door',
      backdrop_path: '/vj4IhmH4HCMZYYjTMiYBybTWR5o.jpg',
      popularity: 22.279864,
      vote_count: 628,
      video: false,
      vote_average: 4.13,
    },
    {
      poster_path: '/vOipe2myi26UDwP978hsYOrnUWC.jpg',
      adult: false,
      overview:
        'An orphan boy is raised in the Jungle with the help of a pack of wolves, a bear and a black panther.',
      release_date: '2016-04-07',
      genre_ids: [12, 18, 14],
      id: 278927,
      original_title: 'The Jungle Book',
      original_language: 'en',
      title: 'The Jungle Book',
      backdrop_path: '/eIOTsGg9FCVrBc4r2nXaV61JF4F.jpg',
      popularity: 21.104822,
      vote_count: 1085,
      video: false,
      vote_average: 6.42,
    },
  ],
  total_results: 19629,
  total_pages: 982,
};

export const movieInfo = {
  adult: false,
  backdrop_path: '/fCayJrkfRaCRCTh8GqN30f8oyQF.jpg',
  belongs_to_collection: null,
  budget: 63000000,
  genres: [
    {
      id: 18,
      name: 'Drama',
    },
  ],
  homepage: '',
  id: 550,
  imdb_id: 'tt0137523',
  original_language: 'en',
  original_title: 'Fight Club',
  overview:
    'A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground "fight clubs" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.',
  popularity: 0.5,
  poster_path: null,
  production_companies: [
    {
      id: 508,
      logo_path: '/7PzJdsLGlR7oW4J0J5Xcd0pHGRg.png',
      name: 'Regency Enterprises',
      origin_country: 'US',
    },
    {
      id: 711,
      logo_path: null,
      name: 'Fox 2000 Pictures',
      origin_country: '',
    },
    {
      id: 20555,
      logo_path: null,
      name: 'Taurus Film',
      origin_country: '',
    },
    {
      id: 54050,
      logo_path: null,
      name: 'Linson Films',
      origin_country: '',
    },
    {
      id: 54051,
      logo_path: null,
      name: 'Atman Entertainment',
      origin_country: '',
    },
    {
      id: 54052,
      logo_path: null,
      name: 'Knickerbocker Films',
      origin_country: '',
    },
    {
      id: 25,
      logo_path: '/qZCc1lty5FzX30aOCVRBLzaVmcp.png',
      name: '20th Century Fox',
      origin_country: 'US',
    },
  ],
  production_countries: [
    {
      iso_3166_1: 'US',
      name: 'United States of America',
    },
  ],
  release_date: '1999-10-12',
  revenue: 100853753,
  runtime: 139,
  spoken_languages: [
    {
      iso_639_1: 'en',
      name: 'English',
    },
  ],
  status: 'Released',
  tagline:
    "How much can you know about yourself if you've never been in a fight?",
  title: 'Fight Club',
  video: false,
  vote_average: 7.8,
  vote_count: 3439,
};
