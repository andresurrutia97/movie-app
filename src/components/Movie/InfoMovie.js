import React from 'react';
import PropTypes from 'prop-types';

import Details from './Details';
import MovieGenres from '../Ui/MovieGenres';
import Overview from './Overview';
import Title from './Title';
import FavoriteToggle from '../Ui/FavoriteToggle';
import { formatYear, formatCountry } from '../../utils/formatData';

const InfoMovie = ({ data, isFav, addFav, removeFav }) => {
  const year = formatYear(data?.release_date);
  const country = formatCountry(data?.production_countries);

  return (
    <div className="w-full flex flex-col items-center px-5 mb-5 md:w-full md:px-20 md:items-start">
      <div className="flex flex-col w-full mt-5 items-center md:flex-row md:justify-between md:items-center md:mt-6">
        <Title title={data.title} />
        <div className=" right-0 mt-1 md:mt-0">
          <FavoriteToggle
            isFav={isFav}
            add={addFav}
            remove={removeFav}
            movieId={data.id}
          />
        </div>
      </div>
      <MovieGenres genres={data.genres} />
      <div className="flex justify-around mb-2 md:justify-start">
        <Details title="Year" name={year} />
        <Details title="Country" name={country} />
        <Details title="Duration" name={`${data.runtime} min`} />
      </div>
      <div className="md:w-6/12">
        <Overview overview={data.overview} />
      </div>
    </div>
  );
};

InfoMovie.propTypes = {
  data: PropTypes.shape({
    id: PropTypes.number,
    release_date: PropTypes.string,
    production_countries: PropTypes.array,
    title: PropTypes.string,
    genres: PropTypes.array,
    runtime: PropTypes.number,
    overview: PropTypes.string,
  }).isRequired,
  isFav: PropTypes.bool,
  addFav: PropTypes.func,
  removeFav: PropTypes.func,
};

InfoMovie.defaultProps = {
  isFav: false,
  addFav: () => {},
  removeFav: () => {},
};

export default InfoMovie;
