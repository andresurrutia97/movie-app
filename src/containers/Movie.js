import React from 'react';

import { useDispatch, useSelector } from '../utils/reduxHooks';
import MovieView from '../views/Movie';
import { fetchMovieInfo } from '../store/actions/MovieActions';
import {
  addFavorite,
  removeFavorite,
  fetchFavoriteList,
} from '../store/actions/FavoritesActions';
import { addSeen, fetchSeenList } from '../store/actions/SeenMoviesActions';
import { historyShape } from '../utils/propTypes';

const Movie = ({ history }) => {
  const { movieInfo, loading, error, favoritesList, seenList } = useSelector(
    (state) => ({
      movieInfo: state.movie.movieInfo,
      loading: state.movie.loading,
      error: state.movie.error,
      favoritesList: state.favorites.favoritesList,
      seenList: state.seen.seenList,
    })
  );

  const dispatch = useDispatch();

  React.useEffect(() => {
    const id = history.location.pathname.split('/')[2];
    dispatch(fetchMovieInfo(id));
    dispatch(fetchFavoriteList());
    dispatch(fetchSeenList());
  }, [dispatch, history]);

  /**
   * handleAddFavorite() dispatch de addFavorite action to add a new movie
   * to favorites
   */
  const handleAddFavorite = (id) => {
    dispatch(addFavorite(id));
  };

  /**
   * handleRemoveFavorite() dispatch de removeFavorite action to add a new movie
   * to favorites
   */
  const handleRemoveFavorite = (id) => {
    dispatch(removeFavorite(id));
  };

  /**
   * handleAddFavorite() dispatch de addFavorite action to add a new movie
   * to favorites
   */
  const handleAddSeen = (id) => {
    dispatch(addSeen(id));
  };

  return (
    <MovieView
      movieInfo={movieInfo}
      loading={loading}
      error={error}
      favList={favoritesList}
      addFav={handleAddFavorite}
      removeFav={handleRemoveFavorite}
      seenList={seenList}
      addSeen={handleAddSeen}
    />
  );
};

Movie.propTypes = {
  history: historyShape.movieId.isRequired,
};

export default Movie;
