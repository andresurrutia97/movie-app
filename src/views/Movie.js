import React from 'react';
import PropTypes from 'prop-types';

import VideoPlayer from '../components/Ui/VideoPlayer';
import InfoMovie from '../components/Movie/InfoMovie';
import ImageCarousel from '../components/Ui/ImageCarousel';
import ImageCard from '../components/Ui/ImageCard';
import Spinner from '../components/Ui/Spinner';
import Error from '../components/Ui/Error';
import { getImageUrl, getTrailerUrl } from '../utils/urls';
import { movieInfoShape, errorShape } from '../utils/propTypes';

const Movie = ({
  movieInfo,
  loading,
  error,
  favList,
  addFav,
  removeFav,
  seenList,
  addSeen,
}) => {
  let movie = <Spinner />;

  React.useEffect(() => {
    if (movieInfo) {
      const isSeen = seenList
        ? Object.keys(seenList).includes(movieInfo.id.toString())
        : false;
      if (!isSeen) addSeen(movieInfo.id);
    }
  }, [movieInfo]);

  if (movieInfo && !loading) {
    const trailerUrl = getTrailerUrl(movieInfo?.videos);
    const isFav = favList
      ? Object.keys(favList).includes(movieInfo.id.toString())
      : false;

    movie = (
      <div className="w-full movie">
        <VideoPlayer url={trailerUrl} />
        <InfoMovie
          data={movieInfo}
          isFav={isFav}
          addFav={addFav}
          removeFav={removeFav}
        />
        {movieInfo.images ? (
          <div className="pl-5 md:pl-20">
            <ImageCarousel>
              {movieInfo.images.backdrops.map((poster) => (
                <ImageCard
                  key={poster.file_path}
                  imgUrl={getImageUrl(poster.file_path)}
                />
              ))}
            </ImageCarousel>
          </div>
        ) : (
          ''
        )}
      </div>
    );
  } else if (error) {
    movie = <Error />;
  }

  return <>{movie}</>;
};

Movie.propTypes = {
  movieInfo: PropTypes.shape(movieInfoShape),
  loading: PropTypes.bool,
  error: PropTypes.shape(errorShape),
  favList: PropTypes.shape({
    id: PropTypes.number,
  }),
  addFav: PropTypes.func,
  removeFav: PropTypes.func,
  seenList: PropTypes.shape({
    id: PropTypes.number,
  }),
  addSeen: PropTypes.func,
};

Movie.defaultProps = {
  movieInfo: null,
  loading: false,
  error: null,
  favList: null,
  addFav: () => {},
  removeFav: () => {},
  seenList: null,
  addSeen: () => {},
};

export default Movie;
