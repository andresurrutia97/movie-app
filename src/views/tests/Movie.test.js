import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Movie from '../Movie';
import { movieInfo } from '../../utils/fakeData';
import Spinner from '../../components/Ui/Spinner';
import Error from '../../components/Ui/Error';
import VideoPlayer from '../../components/Ui/VideoPlayer';
import InfoMovie from '../../components/Movie/InfoMovie';

describe('<Movie/>', () => {
  let wrapper;
  const initProps = {
    movieInfo: null,
    loading: false,
    error: null,
  };

  const component = <Movie {...initProps} />;

  beforeEach(() => {
    wrapper = mount(component);
  });

  it('should render correcty ', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should contain a instance of <Spinner/> while no info ', () => {
    expect(wrapper.find(Spinner)).toHaveLength(1);
  });

  it('should contain an instance of <VideoPlayer/>, <InfoMovie/> if latestMovie prop contains info', () => {
    wrapper.setProps({
      movieInfo: { movieInfo },
    });
    expect(wrapper.find(VideoPlayer)).toHaveLength(1);
    expect(wrapper.find(InfoMovie)).toHaveLength(1);
  });

  it('should contain 1 instance of <Error/> if error prop caintains info ', () => {
    wrapper.setProps({ error: { error: 'error' } });
    expect(wrapper.find(Error)).toHaveLength(1);
  });
});
