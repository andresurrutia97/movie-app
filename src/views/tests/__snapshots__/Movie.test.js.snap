// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movie/> should render correcty  1`] = `
<div
  className="spinner"
/>
`;
